internal/host/host.go

```diff
@@ -419,3 +414,66 @@ func (m *Manager) reportInstallationMetrics(ctx context.Context, h *models.Host,
 		m.metricApi.ReportHostInstallationMetrics(log, cluster.OpenshiftVersion, h, previousProgress, CurrentStage)
 	}
 }
+
+func (m *Manager) AutoAssignRole(ctx context.Context, h *models.Host, db *gorm.DB) error {
+	// select role if needed
+	if h.Role == models.HostRoleAutoAssign {
+		return m.autoRoleSelection(ctx, h)
+	}
+	return nil
+}
+
+func (m *Manager) selectRole(ctx context.Context, h *models.Host) (models.HostRole, error) {
+	var (
+		autoSelectedRole = models.HostRoleWorker
+		log              = logutil.FromContext(ctx, m.log)
+	)
+
+	// count already existing masters
+	mastersCount := 0
+	if err := m.db.Model(&models.Host{}).Where("cluster_id = ? and status != ? and role = ?",
+		h.ClusterID, models.HostStatusDisabled, models.HostRoleMaster).Count(&mastersCount).Error; err != nil {
+		log.WithError(err).Errorf("failed to count masters in cluster %s", h.ClusterID.String())
+		return autoSelectedRole, err
+	}
+
+	if mastersCount < minMasterHosts {
+		h.Role = models.HostRoleMaster
+		vc, err := newValidationContext(h, m.db)
+		if err != nil {
+			log.WithError(err).Errorf("failed to create new validation context for host %s", h.ID.String())
+			return autoSelectedRole, err
+		}
+		conditions, _, err := m.rp.preprocess(vc)
+		if err != nil {
+			log.WithError(err).Errorf("failed to run validations on host %s", h.ID.String())
+			return autoSelectedRole, err
+		}
+		if conditions[HasCPUCoresForRole] && conditions[HasMemoryForRole] {
```

Comment:
It would be cleaner if we could run the code that transitions a host to known status with the role to see that it passes (in case we add a condition for known master later on and forget to add it here).

Reply:
Not sure i understand, role is not the only thing that set the host to known so not sure i understand your suggestion.

Reply:
Sorry I guess I didn't explain it clearly.  Maybe the comment is similar to Ori's about can-be-a-master.  My concern is that we might add another condition for master in the future and we will forget to update the code here.

Reply:
OK now i understand i will add another validator.